fraud/sync.go

```diff
@@ -0,0 +1,84 @@
+package fraud
+
+import (
+	"context"
+
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+
+	pb "github.com/celestiaorg/celestia-node/fraud/pb"
+)
+
+func (f *ProofService) syncFraudProofs(ctx context.Context) {
+	log.Debug("start fetching Fraud Proofs")
+	sub, err := f.host.EventBus().Subscribe(&event.EvtPeerIdentificationCompleted{})
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	defer sub.Close()
+	f.topicsLk.RLock()
+	proofTypes := make([]pb.ProofType, 0, len(f.topics))
+	for proofType := range f.topics {
+		p, err := proofTypeToPb(proofType)
+		if err != nil {
+			log.Warn(err)
+			continue
+		}
+		proofTypes = append(proofTypes, p)
+	}
+	f.topicsLk.RUnlock()
+	connStatus := event.EvtPeerIdentificationCompleted{}
+	for i := 0; i < fraudRequests; i++ {
+		select {
+		case <-ctx.Done():
+			return
+		case e := <-sub.Out():
+			connStatus = e.(event.EvtPeerIdentificationCompleted)
+		}
+
+		if connStatus.Peer == f.host.ID() {
+			i--
+			continue
+		}
+		go func(pid peer.ID) {
+			log.Debug("requesting proofs from the peer ", pid)
+			respProofs, err := requestProofs(ctx, f.host, pid, proofTypes)
+			if err != nil {
+				log.Errorw("error while requesting fraud proofs from the peer", "err", err, "peer", pid)
+				return
+			}
+			if len(respProofs) == 0 {
+				log.Debug("proofs were not found")
+				return
+			}
+			log.Debug("got fraud proofs from the peer: ", connStatus.Peer)
+			for _, data := range respProofs {
+				proofType, err := pbToProofType(data.Type)
+				if err != nil {
+					log.Warn(err)
+					continue
+				}
+				f.topicsLk.RLock()
+				topic, ok := f.topics[proofType]
+				f.topicsLk.RUnlock()
```

Comment:
do we have to keep reading from this topic map on the service? Can't we just copy the map once and refer to that? It's not like we handle for any changes here since we've already requested only the proof types generated on L23

Reply:
https://github.com/celestiaorg/celestia-node/pull/917#discussion_r951302654